test/distributed/_composable/fsdp/test_fully_shard_init.py

```diff
@@ -554,6 +566,91 @@ def test_invalid_meta_device_init(self):
         with self.assertRaisesRegex(RuntimeError, error_regex):
             model(inp)
 
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_rank0_broadcast_meta_device_init(self):
+        model_args = ModelArgs(dropout_p=0.0)
+        # Assume we have a CPU full state dict on rank 0
+        if self.rank == 0:
+            torch.manual_seed(42)
+            ref_model = Transformer(model_args)
+            full_sd = ref_model.state_dict()
+            for param in full_sd.values():
+                self.assertEqual(param.device, torch.device("cpu"))
+
+        # Initialize the sharded model on meta device
+        fsdp_mesh = init_device_mesh("cuda", (self.world_size,))
+        with torch.device("meta"):
+            model = Transformer(model_args)
+        for module in model.modules():
+            if isinstance(module, TransformerBlock):
+                fully_shard(module, mesh=fsdp_mesh)
+        fully_shard(model, mesh=fsdp_mesh)
+        for param in model.parameters():
+            self.assertEqual(param.device, torch.device("meta"))
+
+        # Construct a sharded state dict from the rank 0 full state dict by
+        # broadcasting and sharding
+        meta_sharded_sd = model.state_dict()
+        sharded_sd = {}
+        if self.rank == 0:
```

Comment:
If per-tensor `broadcast` is achieving poor network utilization and is intolerable, then the user/author of this converter can change this to do bucketing with `dist._broadcast_coalesced`. The implementation will be more involved though.

Reply:
In future I think we can also add a dedicated API maybe: `distribute_multi_tensor` that shard a list of tensors with the same placements